README.org
#+TITLE: immersive-translate: Immersive Bilingual Translation (WIP)
* 项目简介
受 [[https://github.com/immersive-translate/immersive-translate][immersive-translate/immersive-translate]] 的启发，在 Emacs 中实现了类似的功能。
* 使用截图
** elfeed
[[images/screenshots/elfeed.png]]
** nov
[[images/screenshots/nov.png]]
** info
[[images/screenshots/info.png]]
** helpful/help
[[images/screenshots/help.png]]
* 安装
您的设备上需要先安装 [[https://curl.se/][curl]] ，如果需要使用 translate-shell 后端，则需要安装
[[https://github.com/soimort/translate-shell][translate-shell]] 。

#+begin_src emacs-lisp
  (add-to-list 'load-path "/path-to/emacs-immersive-translate/")
  (require 'immersive-translate)
  (add-hook 'elfeed-show-mode-hook #'immersive-translate-setup)
  (add-hook 'nov-pre-html-render-hook #'immersive-translate-setup)
  ;; use Baidu Translation
  (setq immersive-translate-backend 'baidu
		immersive-translate-baidu-appid "your-appid")

  ;; use ChatGPT
  ;; (setq immersive-translate-backend 'chatgpt
  ;; 	  immersive-translate-chatgpt-host "api.openai.com")

  ;; use translate-shell
  ;; (setq immersive-translate-backend 'trans)

#+end_src
* 使用
- =immersive-translate-buffer=
  
  翻译当前 buffer 。 *注意*: 这可能会消耗大量 Token 。

- =immersive-translate-paragraph=

  翻译当前段落。

- =immersive-translate-clear=
  
  清除所有翻译文本。

- =immersive-translate-auto-mode=

  开启后，自动翻译当前窗口可见部分的内容。

- =immersive-translate-abort=
  
  停止当前 buffer 所有的正在进行的翻译进程。
* 自定义设置
- =immersive-translate-backend=

  所使用的翻译后端。详见 [[#api_key][API 密钥设置]] 一节。

  目前可用的翻译服务
  - chatgpt (https://openai.com/api/)
  - baidu (https://fanyi-api.baidu.com/)
  - translate-shell (https://github.com/soimort/translate-shell)

- =immersive-translate-auto-idle=

  开启 =immersive-translate-auto-mode= 后，执行下次翻译时所需要的 Emacs 空闲时间。
  
- =immersive-translate-exclude-shr-tag=

  此变量中的标签所包含的文本都不会翻译 。此变量适用于所有由 =shr.el= 生成的 buffer，
  如 =elfeed-show-mode= 、=nov-mode= 和 =mu4e-view-mode= 中的 buffer 。

- =immersive-translate-chatgpt-system-prompt=
  
  System prompt used by ChatGPT.

- =immersive-translate-chatgpt-user-prompt=
  
  User prompt used by ChatGPT.

- =immersive-translate-disable-predicates=
  
  当当前段落不需要翻译时，返回 =t= 。此列表中的函数不接受任何参数。


更多自定义配置见：=M-x customize-group RET immersive-translate RET=
* API 密钥设置
:PROPERTIES:
:CUSTOM_ID: api_key
:END:
此插件的翻译服务需要你申请对应服务的 API 密钥才能使用，详见
  [[https://immersivetranslate.com/docs/services/][翻译服务 API 申请]] 。

此插件通过内置的 =auth-source.el= 来获取用户的密钥：

1. ChatGPT
   
   假设您的 =immersive-translate-chatgpt-host= 为 "api.openai.com" , API 密钥为
   123456 。则在 =~/.authinfo= 中添加如下文本：
  #+begin_example
  machine api.openai.com login apikey password 123456
  #+end_example
  
2. 百度翻译
   
   假设您的 =immersive-translate-baidu-appid= 为 "foobar" ，API 密钥为123456 。则在
   =~/.authinfo= 中添加如下文本：
  #+begin_example
  machine fanyi-api.baidu.com login foobar password 123456
  #+end_example
* 添加新的翻译后端 (欢迎 PR!)
** 添加在线翻译服务
如果想要添加新的翻译服务，可以查看变量
=immersive-translate-curl-get-translation-alist= 和
=immersive-translate-curl-get-args-alist= 。

变量 =immersive-translate-curl-get-args-alist= 中的函数用于构造请求时所需的输入
参数。

变量 =immersive-translate-curl-get-translation-alist= 中的函数用于解析 =curl= 返回的
翻译结果。
** 添加本地翻译软件
参考 [[immersive-translate-trans.el]]


CHANGELOG.org
#+title: Emacs-immersive-translate changelog
* v0.3.0(2023-07-24)
** New feaure: separate front-end and translating backends
- Use =immersive-translate-backend= to select different translation services.
- Use =immersive-translate-backend-alist= to add new backends.
** Add =immersive-translate--format-translation= 
** Add =Translate-shell= backend
* v0.2.0 (2023-07-21)
** New feature: enable multiple translation services
Now users can select different translation services by setting =immersive-translate-service=.
** New feature: add =immersive-translate-auto-mode=


immersive-translate.el
;;; immersive-translate.el --- translate the current buffer immersively -*- lexical-binding: t; -*-

;; Copyright (C) 2023  Eli Qian

;; Author: Eli Qian <eli.q.qian@gmail.com>
;; Url: https://github.com/Elilif/emacs-immersive-translate

;; Version: v0.3.0
;; Keywords: translation
;; SPDX-License-Identifier: GPL-3.0-or-later

;;; Commentary:


;;; Code:

(require 'dom)
(require 'auth-source)
(require 'text-property-search)
(require 'immersive-translate-baidu)
(require 'immersive-translate-chatgpt)
(require 'immersive-translate-trans)



;;;; Customizations
(defgroup immersive-translate nil
  "Immersive translation."
  :group 'applications)

(defcustom immersive-translate-auto-idle 0.5
  "Perform translation the next time Emacs is idle for seconds."
  :group 'immersive-translate
  :type 'number)

(defcustom immersive-translate-backend 'baidu
  "The translation backend to use.

The current options are
- chatgpt
- baidu
- trans"
  :group 'immersive-translate
  :type '(choice
          (const :tag "ChatGPT" chatgpt)
          (const :tag "Baidu" baidu)
          (const :tag "translate-shell" trans)))

(defcustom immersive-translate-backend-alist
  '((baidu . immersive-translate-baidu-translate)
    (chatgpt . immersive-translate-chatgpt-translate)
    (trans . immersive-translate-trans-translate))
  "Alist of functions to do the translation job.

Each element looks like (BACKEND . FUNCTION);

BACKEND is the translation backend, see
`immersive-translate-backend'; FUNCTION is a function of two
argument."
  :group 'immersive-translate
  :type '(alist :key-type symbol :value-type function))

(defcustom immersive-translate-pending-message "🔄"
  "Text displayed before the translation results are returned."
  :group 'immersive-translate
  :type 'string)

(defcustom immersive-translate-failed-message "🔀"
  "Text displayed when translation fails."
  :group 'immersive-translate
  :type 'string)

(defcustom immersive-translate-exclude-shr-tag '(html
                                                 base
                                                 body
                                                 a
                                                 audio
                                                 title
                                                 img
                                                 pre
                                                 script
                                                 style
                                                 svg
                                                 span
                                                 sub
                                                 sup
                                                 br
                                                 code
                                                 tt
                                                 hr
                                                 div
                                                 ol
                                                 ul
                                                 dl
                                                 strong
                                                 i
                                                 u
                                                 em
                                                 table)
  "HTML components that should not be translated."
  :group 'immersive-translate
  :type '(repeat symbol))

(defun immersive-translate--info-code-block-p ()
  "Return non-nil if the current paragraph is a code block."
  (let ((current-line (thing-at-point 'line t)))
    (when (and
           (eq major-mode 'Info-mode)
           (string-match "^\\( \\{5,\\}\\)" current-line))
      (let* ((current-indent (length (match-string 1 current-line)))
             (prev-line (save-excursion
                          (forward-line -2)
                          (thing-at-point 'line t)))
             (prev-indent (if (string-match "^ +" prev-line)
                              (length (match-string 0 prev-line))
                            0))
             (prev-is-kb-desc? (string-prefix-p "‘" prev-line))
             (small-indent-change? (< (- current-indent prev-indent) 5)))
        (or (not (or prev-is-kb-desc? small-indent-change?))
            (string-match-p "^ +\(" current-line))))))

(defun immersive-translate--info-menu-p ()
  "Return non-nil if the current line is a menu."
  (when (eq major-mode 'Info-mode)
    (string-match-p "^\\*" (thing-at-point 'line t))))

(defun immersive-translate--helpful-not-doc-p ()
  "Return non-nil if point is not in the doc section."
  (when (eq major-mode 'helpful-mode)
    (when-let ((doc-beg (save-excursion
                          (goto-char (point-min))
                          (re-search-forward "^Documentation$" nil 'noerror)
                          (line-beginning-position)))
               (doc-end (save-excursion
                          (goto-char (point-min))
                          (re-search-forward "^References$" nil 'noerror)
                          (line-beginning-position))))
      (or (< (point) doc-beg)
          (>= (point) doc-end)))))

(defun immersive-translate--help-signature-p ()
  "Return non-nil if the current paragraph is a signature."
  (when (eq major-mode 'help-mode)
    (string-match-p "^ *\(" (thing-at-point 'line t))))

(defun immersive-translate--translation-exist-p ()
  "Return non-nil if the current paragraph has been translated."
  (save-excursion
    (immersive-translate-end-of-paragraph)
    (when-let ((overlays (overlays-in (1- (point)) (1+ (point)))))
      (cl-some (lambda (ov)
                 (overlay-get ov 'after-string))
               overlays))))

(defcustom immersive-translate-disable-predicates '(immersive-translate--translation-exist-p
                                                    immersive-translate--info-code-block-p
                                                    immersive-translate--info-menu-p
                                                    immersive-translate--helpful-not-doc-p
                                                    immersive-translate--help-signature-p)
  "Predicates, return t when the current paragraph should not be translated.

Predicate functions don't take any arguments."
  :group 'immersive-translate
  :type '(repeat function))

(defvar immersive-translate--process-alist nil
  "Alist of active immersive-translate curl requests.")
(defvar-local immersive-translate--translation-overlays nil)
(defvar immersive-translate--timer nil)
(defvar-local immersive-translate--window-start nil)
(defvar-local immersive-translate--window-end nil)

(defun immersive-translate-api-key (host user)
  "Lookup api key in the auth source.

By default, `immersive-translate-chatgpt-host' is used as HOST
and \"apikey\" as USER in Chatgpt backend.
\"fanyi-api.baidu.com\" is used as HOST and
`immersive-translate-baidu-appid' as USER in Baidu backend."
  (if-let ((secret (plist-get (car (auth-source-search
                                    :host host
                                    :user user
                                    :require '(:secret)))
                              :secret)))
      (if (functionp secret)
          (encode-coding-string (funcall secret) 'utf-8)
        secret)
    (user-error (format "No %s found in the auth source" user))))


;;;; shr
(defun immersive-translate--shr-set-bound (orig dom)
  "Set DOM bound.

Advice for ORIG (shr-tag-*)."
  (let ((beg (point)))
    (funcall orig dom)
    (unless (= beg (point))
      (put-text-property beg (1+ beg) 'immersive-translate--beg (dom-tag dom))
      (save-excursion
        (save-restriction
          (widen)
          (skip-chars-backward "[ \n\t]")
          (put-text-property (1- (point)) (point) 'immersive-translate--end (dom-tag dom)))))))

(defun immersive-translate--shr-tag-advice (dom)
  "Add advice for DOM."
  (let* ((tag (dom-tag dom))
         (function
          (intern (concat "shr-tag-" (symbol-name tag)) obarray)))
    (unless (memq tag immersive-translate-exclude-shr-tag)
      (advice-add function :around #'immersive-translate--shr-set-bound))))


;;;; paragraph get functions
(defun immersive-translate--info-get-paragraph ()
  "Return the paragraph at point."
  (forward-paragraph -1)
  (forward-line)
  (let* ((eop (save-excursion
                (end-of-paragraph-text)
                (point)))
         (bop (save-excursion
                (re-search-forward "\\(^ [^ ].*\n \\{5,\\}\\)\\|\\(^ *‘.*’$\\)"
                                   (line-end-position 2)
                                   'noerror)))
         (string (if bop
                     (buffer-substring-no-properties bop eop)
                   (buffer-substring-no-properties (point) eop))))
    (thread-last
      string
      (replace-regexp-in-string "\n" " ")
      (replace-regexp-in-string " +" " "))))

(defun immersive-translate--helpful-get-paragraph ()
  "Return the paragraph at point."
  (forward-paragraph -1)
  (forward-line)
  (let ((para (thing-at-point 'paragraph t)))
    (string-trim-left para "\nDocumentation\n")))

(defun immersive-translate--elfeed-get-paragraph ()
  "Return the paragraph at point."
  (save-excursion
    (immersive-translate-end-of-paragraph)
    (when-let* ((beg-prop (text-property-search-backward 'immersive-translate--beg))
                (beg-pos (prop-match-beginning beg-prop))
                (tag (prop-match-value beg-prop))
                (end-prop (text-property-search-forward 'immersive-translate--end))
                (end-pos (prop-match-beginning end-prop)))
      (if (eq tag 'li)
          (buffer-substring-no-properties (+ beg-pos 2) end-pos)
        (buffer-substring-no-properties beg-pos end-pos)))))

(defun immersive-translate--elfeed-tube-get-paragraph ()
  "Return the paragraph at point."
  (forward-paragraph -1)
  (forward-line)
  (let ((para (thing-at-point 'paragraph t)))
    (string-trim-left para "\n\\[.*?\\] \\- \\[.*?\\]:\n")))

(defun immersive-translate--elfeed-tube-p (&optional mode)
  "Return non-nil if the current feed in MODE is a Youtube RSS feed.

Return nil otherwise."
  (when (eq mode 'elfeed-show-mode)
    (save-excursion
      (goto-char (point-min))
      (text-property-search-forward 'timestamp))))

(defun immersive-translate--get-paragraph ()
  "Return the paragraph at point."
  (pcase major-mode
    ('Info-mode
     (immersive-translate--info-get-paragraph))
    ('helpful-mode
     (immersive-translate--helpful-get-paragraph))
    ((pred immersive-translate--elfeed-tube-p)
     (immersive-translate--elfeed-tube-get-paragraph))
    ((or 'elfeed-show-mode 'nov-mode 'mu4e-view-mode)
     (immersive-translate--elfeed-get-paragraph))
    (_
     (thing-at-point 'paragraph t))))


;;;; format functions
(defun immersive-translate--get-indent (marker)
  "Return the indentation of the original text at MARKER."
  (with-current-buffer (marker-buffer marker)
    (save-excursion
      (goto-char marker)
      (save-excursion
        (re-search-backward "^\\( *\\)" (line-beginning-position))
        (match-string 1)))))

(defun immersive-translate--format-translation (str marker)
  "Function which produces the string to insert as a translation.

STR is the original translation. MARKER is the position where the
translation should be inserted."
  (with-temp-buffer
    (insert str)
    (fill-region-as-paragraph (point-min) (point-max))
    (concat
     "\n"
     (replace-regexp-in-string
      "^"
      (immersive-translate--get-indent marker)
      (buffer-substring-no-properties
       (point-min) (point-max)))
     "\n")))

(defun immersive-translate--info-transform-response (str marker)
  "Format STR at MARKER in `Info-mode'."
  (let* ((fill-column 70)
         (str (string-trim-right str "[-=]+")))
    (immersive-translate--format-translation str marker)))

(defun immersive-translate--nov-transform-response (str marker)
  "Format STR at MARKER in `nov-mode'."
  (let ((fill-column (or (and (boundp 'nov-text-width)
                              nov-text-width)
                         120)))
    (immersive-translate--format-translation str marker)))

(defun immersive-translate--elfeed-transform-response (str marker)
  "Format STR at MARKER in `elfeed-mode'."
  (let ((fill-column (or (and (boundp 'shr-width)
                              shr-width)
                         110)))
    (save-excursion
      (goto-char marker)
      (let* ((beg-prop (text-property-search-backward 'immersive-translate--beg))
             (tag (prop-match-value beg-prop)))
        (concat (immersive-translate--format-translation str marker)
                (if (eq tag 'li)
                    "\n"
                  ""))))))

(defun immersive-translate--help-transform-response (str marker)
  "Format STR at MARKER in `help-mode'."
  (let ((fill-column 70))
    (immersive-translate--format-translation str marker)))

(defun immersive-translate--transform-response (content-str &optional marker)
  "Format CONTENT-STR at MARKER."
  (pcase major-mode
    ('Info-mode
     (immersive-translate--info-transform-response content-str marker))
    ('nov-mode
     (immersive-translate--nov-transform-response content-str marker))
    ((or 'elfeed-show-mode 'mu4e-view-mode)
     (immersive-translate--elfeed-transform-response content-str marker))
    ((or 'helpful-mode
         'help-mode)
     (immersive-translate--help-transform-response content-str marker))
    (_
     (immersive-translate--format-translation content-str marker))))


;;;; utility functions
(defun immersive-translate-callback (response info)
  "Insert RESPONSE from ChatGPT into the current buffer.

INFO is a plist containing information relevant to this buffer."
  (when-let* (((not (string-empty-p response)))
              (status-str  (plist-get info :status))
              (origin-buffer (plist-get info :buffer))
              (start-marker (plist-get info :position)))
    (with-current-buffer origin-buffer
      (if response
          (progn
            (setq response (immersive-translate--transform-response response start-marker))
            (save-excursion
              (with-current-buffer (marker-buffer start-marker)
                (goto-char start-marker)
                (let ((ovs (overlays-in (1- (point)) (point)))
                      (new-ov (make-overlay (point) (1+ (point)))))
                  (mapc (lambda (ov)
                          (when (overlay-get ov 'after-string)
                            (delete-overlay ov)))
                        ovs)
                  (overlay-put new-ov
                               'after-string
                               response)
                  (push new-ov immersive-translate--translation-overlays)))))
        (message "Response error: (%s) %s"
                 status-str (plist-get info :error))))))

(defun immersive-translate-disable-p ()
  "Return non-nil if the current paragraph should not be translated.

Nil otherwise."
  (cl-some (lambda (pred)
             (funcall pred))
           immersive-translate-disable-predicates))

(defun immersive-translate-end-of-paragraph ()
  "Move to the end of the current paragraph."
  (pcase major-mode
    ((and (or 'elfeed-show-mode 'nov-mode 'mu4e-view-mode)
          (pred (not immersive-translate--elfeed-tube-p)))
     (unless (get-text-property (point) 'immersive-translate--beg)
       (text-property-search-backward 'immersive-translate--beg))
     (text-property-search-forward 'immersive-translate--end)
     (end-of-line))
    (_ (end-of-paragraph-text))))

(defun immersive-translate-join-lin (paragraph)
  "Form the current PARAGRAPH into a continuous paragraph."
  (when paragraph
    (let ((new-str (string-clean-whitespace
                    (replace-regexp-in-string "\n" " " paragraph))))
      (unless (string-empty-p new-str)
        new-str))))

(defun immersive-translate-region (start end)
  "Translate the text between START and END."
  (save-excursion
    (goto-char start)
    (pcase major-mode
      ((and (or 'elfeed-show-mode 'nov-mode 'mu4e-view-mode)
            (pred (not immersive-translate--elfeed-tube-p)))
       (while (and (text-property-search-forward 'immersive-translate--end)
                   (< (point) end))
         (immersive-translate-paragraph)))
      (_ (while (and
                 (< (point) end)
                 (re-search-forward "^\\s-*$" end 'noerror)
                 (not (eobp)))
           (forward-line)
           (immersive-translate-paragraph))))))

;;;###autoload
(defun immersive-translate-abort (buf)
  "Stop all active immersive-translate processes in BUF."
  (interactive (list (current-buffer)))
  (if-let* ((proc-attrs
             (cl-remove-if-not
              (lambda (proc-list)
                (eq (plist-get (cdr proc-list) :buffer) buf))
              immersive-translate--process-alist)))
      (dolist (proc-attr proc-attrs)
        (let ((proc (car proc-attr)))
          (setf (alist-get proc immersive-translate--process-alist nil 'remove) nil)
          (set-process-sentinel proc #'ignore)
          (delete-process proc)
          (kill-buffer (process-buffer proc))
          (message "Stopped all immersive-translate processes in buffer %S" (buffer-name buf))))
    (message "No immersive-translate process associated with buffer %S" (buffer-name buf))))

(cl-defun immersive-translate-do-translate
    (&optional content &key callback
               (buffer (current-buffer))
               position
               (backend immersive-translate-backend))
  "Request a response from `immersive-translate-backend' for CONTENT.

If PROMPT is
- a string, it is used as is
- A list of plists, it is used to create a full prompt suitable for
  sending to ChatGPT.

Keyword arguments:

CALLBACK, if supplied, is a function of two arguments, called
with the RESPONSE (a string) and INFO (a plist):

\(callback RESPONSE INFO)

RESPONSE is nil if there was no response or an error.

The INFO plist has (at least) the following keys:
:content       - The content that was sent with the request
:buffer       - The buffer current when the request was sent.
:position     - marker at the point the request was sent.
:status       - Short string describing the result of the request

Example of a callback that messages the user with the response
and info:

\(lambda (response info)
  (if response
      (let ((posn (marker-position (plist-get info :position)))
            (buf  (buffer-name (plist-get info :buffer))))
        (message \"Response for request from %S at %d: %s\"
                 buf posn response))
    (message \"request failed with message: %s\"
             (plist-get info :status))))

Or, for just the response:

\(lambda (response _)
  ;; Do something with response
  (message (rot13-string response)))

If CALLBACK is omitted, the response is inserted at the point the
request was sent.

BUFFER is the buffer the request belongs to. If omitted the
current buffer is recorded.

POSITION is a buffer position (integer or marker). If omitted,
the value of (point) is recorded.

BACKEND is the translation backend to use. If omitted, the value
of `immersive-translate-backend' is used."
  (let* ((start-marker
          (cond
           ((null position)
            (if (use-region-p)
                (set-marker (make-marker) (region-end))
              (point-marker)))
           ((markerp position) position)
           ((integerp position)
            (set-marker (make-marker) position buffer))))
         (info (list :content content
                     :buffer buffer
                     :position start-marker)))
    (funcall (alist-get backend immersive-translate-backend-alist) info callback)))

;;;###autoload
(defun immersive-translate-buffer ()
  "Translate the whole buffer."
  (interactive)
  (pcase major-mode
    ((pred immersive-translate--elfeed-tube-p)
     (save-excursion
       (goto-char (point-min))
       (let ((beg (search-forward "Transcript:\n" (point-max) 'noerror)))
         (immersive-translate-region beg (point-max)))))
    (_
     (immersive-translate-region (point-min) (point-max)))))

;;;###autoload
(defun immersive-translate-paragraph ()
  "Translate the current paragraph."
  (interactive)
  (save-excursion
    (unless (immersive-translate-disable-p)
      (when-let* ((paragraph (immersive-translate-join-lin
                              (immersive-translate--get-paragraph)))
                  (content (if (eq immersive-translate-backend 'chatgpt)
                               (immersive-translate-chatgpt-create-prompt paragraph)
                             paragraph))
                  (ov t))
        (immersive-translate-end-of-paragraph)
        (setq ov (make-overlay (1- (point)) (point)))
        (overlay-put ov
                     'after-string
                     immersive-translate-pending-message)
        (immersive-translate-do-translate content)))))

;;;###autoload
(defun immersive-translate-clear ()
  "Clear translations."
  (interactive)
  (let ((ovs (overlays-in (point-min) (point-max))))
    (mapc (lambda (ov)
            (when (overlay-get ov 'after-string)
              (delete-overlay ov)))
          ovs))
  (dolist (ov immersive-translate--translation-overlays)
    (delete-overlay ov))
  (setq immersive-translate--translation-overlays nil))

(defun immersive-translate--auto-translate-window ()
  "Translate the text in the current window."
  (let ((start (window-start))
        (end (window-end)))
    (setq immersive-translate--window-start start
          immersive-translate--window-end end)
    (immersive-translate-region start end)))

;;;###autoload
(define-minor-mode immersive-translate-auto-mode
  "Toggle `immersive-translate-auto-mode'.

Translate paragraph under the cursor after Emacs is idle for
`immersive-translate-auto-idle' seconds."
  :global nil
  :group 'immersive-translate
  (cond
   (immersive-translate-auto-mode
    (when (and immersive-translate--timer (timerp immersive-translate--timer))
      (cancel-timer immersive-translate--timer))
    (immersive-translate--auto-translate-window)
    (setq immersive-translate--timer
          (run-with-idle-timer immersive-translate-auto-idle 'repeat #'immersive-translate--auto-translate)))
   (t
    (cancel-timer immersive-translate--timer)
    (setq immersive-translate--timer nil
          immersive-translate--window-start nil
          immersive-translate--window-end nil))))

(defun immersive-translate--auto-translate ()
  "Helper Function for `immersive-translate-auto-mode'."
  (when (and immersive-translate-auto-mode
             immersive-translate--window-end
             (or (< (window-start) immersive-translate--window-start)
                 (> (window-end) immersive-translate--window-end)))
    (immersive-translate--auto-translate-window)))

;;;###autoload
(defun immersive-translate-setup ()
  "Setup immersive-trasnlate."
  (advice-add 'shr-descend :before #'immersive-translate--shr-tag-advice))


;;;; provide
(provide 'immersive-translate)
;;; immersive-translate.el ends here


immersive-translate-chatgpt.el
;;; immersive-translate-chatgpt.el --- chatgpt backend for immersive-translation -*- lexical-binding: t; -*-

;; Based on work by
;; Copyright (C) 2023  Karthik Chikmagalur
;; Original Author: Karthik Chikmagalur;; <karthikchikmagalur@gmail.com>
;; Keywords: convenience

;; SPDX-License-Identifier: GPL-3.0-or-later

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <https://www.gnu.org/licenses/>.


;;; Commentary:
;; 2023/07/21
;; Most of the code in this file is taken from
;; https://github.com/karthink/gptel/blob/master/gptel-curl.el, with the
;; following modifications:
;; 1. Renamed function and variable names.
;; 3. Modified part of the code to fit this project.


;;; Code:

(require 'immersive-translate-curl)

(declare-function immersive-translate-api-key "ext:immersive-translate")

(defgroup immersive-translate-chatgpt nil
  "Immersive translate chatgpt backend."
  :group 'immersive-translate)

(defcustom immersive-translate-chatgpt-host "api.openai.com"
  "The OpenAI API host queried by immersive-translate."
  :group 'immersive-translate-chatgpt
  :type 'string)

(defcustom immersive-translate-chatgpt-model "gpt-3.5-turbo-0613"
  "GPT Model for chat.

The current options are
- \"gpt-3.5-turbo\"
- \"gpt-3.5-turbo-16k\"
- \"gpt-4\" (experimental)
- \"gpt-4-32k\" (experimental)"
  :group 'immersive-translate-chatgpt
  :type '(choice
          (const :tag "GPT 3.5 turbo" "gpt-3.5-turbo")
          (const :tag "GPT 3.5 turbo 16k" "gpt-3.5-turbo-16k")
          (const :tag "GPT 4 (experimental)" "gpt-4")
          (const :tag "GPT 4 32k (experimental)" "gpt-4-32k")))

(defcustom immersive-translate-chatgpt-temperature 0.7
  "\"Temperature\" of ChatGPT response.

This is a number between 0.0 and 2.0 that controls the randomness
of the response, with 2.0 being the most random."
  :group 'immersive-translate-chatgpt
  :type 'number)

(defcustom immersive-translate-chatgpt-proxy ""
  "Path to a proxy to use for ChatGPT query.
Passed to curl via --proxy arg, for example \"proxy.yourorg.com:80\"
Leave it empty if you don't use a proxy."
  :group 'immersive-translate-chatgpt
  :type 'string)

(defcustom immersive-translate-chatgpt-system-prompt "You are a professional translator."
  "System prompt used by ChatGPT."
  :group 'immersive-translate-chatgpt
  :type 'string)

(defcustom immersive-translate-chatgpt-user-prompt "You will be provided with text delimited by triple backticks, your task is to translate the wrapped text into Chinese. You should only output the translated text. \n```%s```"
  "User prompt used by ChatGPT."
  :group 'immersive-translate-chatgpt
  :type 'string)

(defun immersive-translate-chatgpt--request-data (prompts)
  "Encode data into JSON format..

Argument PROMPTS are for sending to chatgpt."
  (let ((prompts-plist
         `(:model ,immersive-translate-chatgpt-model
                  :messages [,@prompts])))
    (plist-put prompts-plist :temperature immersive-translate-chatgpt-temperature)
    prompts-plist))

(defun immersive-translate-chatgpt-get-args (prompts token)
  "Produce list of arguments for calling Curl.

PROMPTS is the data to send, TOKEN is a unique identifier."
  (let* ((url (format "https://%s/v1/chat/completions"
                      immersive-translate-chatgpt-host))
         (data (encode-coding-string
                (json-encode (immersive-translate-chatgpt--request-data prompts))
                'utf-8))
         (headers
          `(("Content-Type" . "application/json")
            ("Authorization" . ,(concat "Bearer " (immersive-translate-api-key
                                                   immersive-translate-chatgpt-host
                                                   "apikey"))))))
    (append
     (list "--location" "--silent" "--compressed" "--disable"
           (format "-X%s" "POST")
           (format "-w(%s . %%{size_header})" token)
           (format "-m%s" 60)
           "-D-"
           (format "-d%s" data))
     (when (not (string-empty-p immersive-translate-chatgpt-proxy))
       (list "--proxy" immersive-translate-chatgpt-proxy
             "--proxy-negotiate"
             "--proxy-user" ":"))
     (cl-loop for (key . val) in headers
              collect (format "-H%s: %s" key val))
     (list url))))

(defun immersive-translate-chatgpt-create-prompt (content)
  "Create a full prompt suitable for sending to ChatGPT.

CONTENT is the text to be translated."
  (let ((user-prompt (format immersive-translate-chatgpt-user-prompt content)))
    `((:role "system" :content ,immersive-translate-chatgpt-system-prompt)
      (:role "user"   :content ,user-prompt))))

(defun immersive-translate-curl-chatgpt-get-translation (response)
  "Get the translated text in RESPONSE returned by CHATGPT."
  (map-nested-elt response '(:choices 0 :message :content)))

(add-to-list 'immersive-translate-curl-get-translation-alist
             '(chatgpt . immersive-translate-curl-chatgpt-get-translation))

(add-to-list 'immersive-translate-curl-get-args-alist
             '(chatgpt . immersive-translate-chatgpt-get-args))

(defun immersive-translate-chatgpt-translate (info &optional callback)
  "Translate the content in INFO using ChatGPT.

INFO is a plist with the following keys:
- :content (the text needed to be translated)
- :buffer (the current buffer)
- :position (marker at which to insert the response).

Call CALLBACK with the response and INFO afterwards. If omitted
the response is inserted into the current buffer after point."
  (immersive-translate-curl-do 'chatgpt info callback))

(provide 'immersive-translate-chatgpt)
;;; immersive-translate-chatgpt.el ends here
